enhancements/installer/packet-ipi.md

```diff
@@ -0,0 +1,197 @@
+that is very similar to UPI but simplifies the process.
+
+Following other IPI installers, this first step would include
+
+- Making Packet IPI documentation available here:
+  <https://github.com/openshift/installer/blob/master/docs/user/packet/install_ipi.md>
+- Adapt Packet's sig-lifecyclce ClusterAPI v1alpha2 for use as an OpenShift's ClusterAPI v1beta1 Machine driver
+- Prepare the Terraform code and Packet types necessary for an IPI installer
+- Making a CI job executing the provisioning scripts to test the IPI installer
+
+### Non-Goals
+
+It is outside the scope of this enhancement to provide explanations about the
+installation of infrastructure elements that are considered as required and
+owned by the user (e.g. DNS, Load Balancer).
+
+Additional enhancement requests will be created to determine the best approach for fulfilling each need.
+
+## Proposal
+
+- Define and implement Packet types in the openshift/installer
+  - High level variables should include:
+    - API Key
+    - Project ID
+  - Boostrap node variables include:
+    - Device plan (defaulting to minimum required)
+    - Facility
+    - Virtual Networks
+  - Control plane variables rely on boostrap variables and add:
+    - Device plan (defaulting to provide a good experience)
+  - Additional clusters creation should include
+    - Device plan (defaulting to provide a good experience)
+    - Facility
+    - Virtual Networks
+    - Future: IP Reservations, Hardware Reservations, Spot Market Requests
+- Write documentation to help users use and understand the Packet installer, to include:
+  - Packet usage basics (accounts and API keys)
+  - Packet resource requirements and options
+  - OpenShift Installer options
+  - Non-Packet components that build on the experience (DNS configuration)
+- Setup the CI job to have a running test suite
+
+### Implementation Details/Notes/Constraints
+
+- RHCOS must be made available and maintained.
+  Images available for provisioning may be custom or official.
+  Official images can be hosted outside of Packet, such as ContainerLinux and RancherOS. More details area available here: <https://github.com/packethost/packet-images>
+  Custom Images can be hosted on git, as described here:
+  - <https://www.packet.com/developers/docs/servers/operating-systems/custom-images/>
+  - <https://www.packet.com/developers/changelog/custom-image-handling/>
+- Packet does not offer a managed DNS solution. Route53, CloudFlare, or RFC 2136
+  (DNS Update), and NS record based zone forwarding are considerations. Existing
+  IPI drivers like BareMetal have similar limitations.
+- LoadBalancing can be achieved with the Packet CCM, which creates and maintains
+  a MetalLB deployment within the cluster.
+- Both Public and Private IPv4 and IPv6 addresses made available for each node.
```

Comment:
We should discuss how network configuration is to be handled.  IIRC, the IPv4 address is handed out via DHCP, so that should work fine without any further changes.  There is no DHCPv6 for IPv6, so OpenShift configured for IPv6 will not work without some additional changes.  I believe we need the hosts to pull their networking config from a packet API so they can self-configure the IPv6 address?

Reply:
Packet devices do not use DHCP.  Device addresses are configured on the filesystem early in the boot phase.
We can look at how this is implemented to some degree with the opensource Tinkerbell project:

- https://github.com/tinkerbell/osie/blob/master/ci/network-test-files/want/centos-x1.small.x86-supermicro/etc/sysconfig/network-scripts/ifcfg-bond0https://github.com/tinkerbell/osie/blob/96d01cf72d64dc0b0d32e74631a467af64d4c4ad/docker/scripts/osie.sh#L551
- https://github.com/packethost/packet-networking/tree/master/packetnetworking/distros/redhat
- https://github.com/packethost/packet-images
  (The images are based on and distributed as Docker images. A different approach is taken for images that require raw partitiion/filesystem layouts, as RHCOS does.  This will require some RedHat <-> Packet coordination.)

All of this is to say, that the Packet approach to providing images and preparing those images on boot-up could be explored.

>  believe we need the hosts to pull their networking config from a packet API so they can self-configure the IPv6 address?

There is also a metadata service available.  https://www.packet.com/developers/docs/servers/key-features/metadata/
OSIE configures Cloudinit to use EC2 Data sources with the metadata.packet.net/metadata URL.  Would it benefit this effort to seek Packet addition to https://github.com/canonical/cloud-init/tree/master/cloudinit/sources, or is there a different project where this would need to be added (Ignition)?

It may also be possible to boot a raw unmodified RHCOS image, but I'm not sure about the process.  If this direction were pursued, I assume we would need a Packet presence in the version of CloudInit/Ignition shipped by RHCOS.

Reply:
Thanks for the clarifications!

cloud-init is not used here.  Ignition is used, instead.

I do think we should aim for booting an unmodified RHCOS image and allowing it to get initialized at boot time.  By default we'll be assuming DHCP is available.  If not, it does sound like we'll need to provide some additional ignition configuration to configure networking on first boot.

You may find this enhancement interesting, which describes the levers we have available for configuring networking on the host: https://github.com/openshift/enhancements/pull/394

At some point I'd like to see a section in this document that describes how we'll handle host network configuration for this platform.  While the enhancement is `provisional`, it can be captured as an open question while we work through the details.

Reply:
actually, how does host network configuration work for the UPI installation method already implemented?

Reply:
A bastion node is created to serve the RHCOS image and act as a DHCP server for other nodes that will be added.

The control plane nodes added after the bastion node get [DHCP/ipxe instructions from the bastion node](https://github.com/RedHatSI/openshift-packet-deploy/blob/master/terraform/modules/bastion/templates/ipxe.tpl#L6-L7).  Their DHCP boot info includes ipxe instructions to boot from the bastion node's hosted copy of the RHCOS image.  

That RHCOS image is pulled at provision time from a static address:
* https://github.com/RedHatSI/openshift-packet-deploy/blob/b8ea976cd83cb850b8732fdabeeb42ba85020bf3/terraform/modules/bastion/main.tf#L53-L59
* https://github.com/RedHatSI/openshift-packet-deploy/blob/b8ea976cd83cb850b8732fdabeeb42ba85020bf3/terraform/modules/bastion/main.tf#L91-L111.

Reply:
In the old CoreOS Container Linux images for Packet, Afterburn (then called coreos-metadata) ran in the initramfs, queried the Packet metadata service for IPv6 and private IPv4 addresses and NIC bonding configuration, and wrote out configuration files for networkd.  That code still exists in RHCOS but would need to be updated for NetworkManager: https://github.com/coreos/fedora-coreos-tracker/issues/111

Reply:
I've been experimenting with RHCOS boots via IPXE URL: https://gist.github.com/displague/5282172449a83c7b83821f8f8333a072

In my latest efforts, the serial console (SOS) doesn't appear to come up after reboot.

Reply:
Yeah, kernel args aren't persisted. I think there's a recent feature to be able to persist them. @yuqi-zhang may know.

Reply:
There is, but it's not exposed on the kernel command line, so you'd have to [override the install unit](https://github.com/coreos/coreos-installer/blob/master/doc/customizing-install.md#custom-coreos-installer-invocation).  I suspect the RHCOS GRUB config should be changed to automatically configure the correct serial console on Packet.

Reply:
I don't think `persistent-kargs` is a flag: https://github.com/coreos/coreos-installer#kernel-command-line-options-for-coreos-installer-running-as-a-service

I think kargs have been persisted by default since 4.2? but it's been awhile since I've checked, so I'd defer to Benjamin's more recent knowledge on how to use this

Reply:
Certain kargs are automatically whitelisted.  `console` isn't one of them, so you'd need to use `--append-karg`.

Reply:
@bgilbert is `append-karg` available in the kernel command line?

In this openshift-installer I think it is fair to have a very opinionated (generated) Ignition file.

For my local testing, and in the general use-cases of launching RHCOS (and related OSes), it would be very helpful (a requirement?) to not need to merge values into the data stashed at the Ignition URL.  I suspect the format of this content could be unpredictable (json vs yaml, other?) (I'm also thinking of cloud-init, where the value could be a shell script or a cloud-config, a gzipped multi-part mime document, or other).

Reply:
> is append-karg available in the kernel command line?

Nope.

On 4.6, with Ignition spec 3, you might take a look at [FCCT](https://github.com/coreos/fcct/), particularly [`storage.trees`](https://github.com/coreos/fcct/blob/master/docs/configuration-v1_1.md).  You can create an FCC that points to a local directory tree, then run FCCT against it to produce an Ignition config which embeds 1) the contents of that tree, and 2) the Ignition config generated by openshift-installer.

Reply:
@bgilbert is FCCT the preferred (or only) method to append kargs? [MachineConfigs](https://github.com/openshift/installer/blob/master/docs/user/customization.md#nodes-with-custom-kernel-arguments) looked like a possible alternative, but I wasn't certain if the kargs whitelist applies here.

Reply:
@liveaverage FCCT can't append kargs right now; that paragraph was replying to the last paragraph of https://github.com/openshift/enhancements/pull/436#discussion_r477349794.  The mechanism you linked is separate from the coreos-installer one, and the whitelist doesn't apply there.  AIUI the MachineConfig karg mechanism doesn't apply to the very first boot.